_tasks/sync-proposal-data.mjs

```diff
@@ -0,0 +1,346 @@
+
+process.on('uncaughtException', (err /* , origin */) => {
+  const errorText = `Uncaught ${
+    err instanceof TypeError ? 'TypeError' : 'Error'
+  }: ${err.message}`;
+
+  writeSync(process.stderr.fd, UnicodeEscapes.errorSymbol + ' ' + errorText +
+            newlineMarker);
+});
+
+// -----------------------------------------------------------------------------
+// Typedefs
+// -----------------------------------------------------------------------------
+
+/**
+ * @typedef {{
+ *   date: string,
+ *   url: string,
+ * }}
+ */
+const PresentedObj = {};
+
+// -----------------------------------------------------------------------------
+// Helpers
+// -----------------------------------------------------------------------------
+
+/**
+ * Gets the URL value from the Markdown link key in the proposals README.
+ * @param {string} urlKeyLike
+ * @returns {string}
+ */
+function getUrlFromDoc(urlKeyLike) {
+  // log.debug({ urlKeyLike });
+  const urlKeyLikeRegex = new RegExp(`\\[${urlKeyLike}.*\\]: (.*)`, 'g');
+  // log.debug({ urlKeyLikeRegex });
+  const results = urlKeyLikeRegex.exec(proposalsReadmeMarkdown);
+  // log.debug({ results });
+  return results[1];
+}
+
+/**
+ * Downloads the proposal's README file using the GitHub API, parses the
+ * Markdown file, and returns the last code block.
+ * @param {string} prpslId
+ * @returns {Array<string>}
+ */
+async function getCodeSample(prpslId) {
+  const codeBlocks = [];
+  try {
+    // Some proposal repos use a different branch for the README.md file (i.e.,
+    // `gh-pages`). Because of this, we need to get some repo metadata before
+    // knowing what the URL is.
+    let rawReadmeUrl;
+    await ghFileImporter
+      .fetchPathMetadata('tc39', prpslId, 'README.md')
+      .then(function (res) {
+        // log.info(res.data.download_url);
+        rawReadmeUrl = res.data.download_url;
+      })
+      .catch(function (err) {
+        throw err;
+      });
+
+    // Get the file and parse the code samples.
+    await ghFileImporter
+      .importFileFromUrl(rawReadmeUrl)
+      .then(async function (res) {
+        const docSourceText = res;
+        await unified()
+          .use(markdown)
+          .use(function () {
+            return function (tree) {
+              visit(tree, null, (node, index) => {
+                if (node.type === 'code') {
+                  if (
+                    node.lang === 'js' ||
+                    node.lang === 'mjs' ||
+                    node.lang === 'javascript'
+                  ) {
+                    codeBlocks.push(he.encode(node.value));
+                  }
+                }
+              });
+            };
+          })
+          .use(remark2rehype, { allowDangerousHtml: true })
+          .use(raw)
+          .use(htmlStringify)
+          .process(docSourceText);
+      })
+      .catch(function (err) {
+        throw err;
+      });
+  } catch (err) {
+    log.error(err);
```

Comment:
do you want to return codeBlocks in this case? I assume not, in which case you’d want:
```suggestion
    log.error(err);
    throw err;
```

Reply:
I see what you mean. If something has gone very wrong in this try block, we probably want to fail early. On the other hand, if the reason for the failure is due to the crawler not being able to find any matching code blocks in the document that it has parsed, I would indeed expect for `codeBlocks` to be returned as an empty array.

Reply:
Sorry, this was probably tough to understand since not only was the type annotation wrong, but the function name is singular too. This bit of the crawl (the code samples) could use some better-defined success and failure criteria so as to get a clearer picture of what all the states it can end up in. For the case of the crawler not finding any code blocks on the page, this logic is described in https://github.com/tc39/tc39.github.io/pull/211/commits/127e90bf2ceac50ed6a6b78b901bbd22de1ee0c7#r488144666, but now that I think about it, this function should probably be returning a single sample string as opposed to an array of all the samples it collected.

This async function (`getCodeSample`) may also wind up throwing or rejecting simply due to situations caused by that third-party Markdown parser that it is using &mdash; might be useful to know what kinds of failures we can expect from it to know whether to anticipate uncaught exceptions, unhandled rejections, etc.

Reply:
Leaving a note here about this since I am still quite bad at this and I don't want to forget. :)

---

> Best Practice – Never Catch for the Purpose of Logging and Then Re-throwing
>
> When trying to figure out why your application is crashing, do not both log the exception and then immediate re-throw it:
>
> ```js
> function clogMyLogs() {
>    try {
>        throw new Error('my error');
>    }
>    catch(e) {
>        console.log(err);
>        throw err;
>    }
> }
> ```
>
> Doing both is redundant and will result in multiple log messages that will clog your logs with the amount of text.
> &mdash; https://joefallon.net/2018/09/typescript-try-catch-finally-and-custom-errors/

---

```ruby
(6:25:22 PM) DerekNonGeneric: is that really a best practice? I hate it
(6:28:45 PM) GreenJello: DerekNonGeneric, it's not really a "never" thing, but rather "don't do this mindlessly"
(6:29:31 PM) GreenJello: some people just log and throw instead of thinking about how to handle errors
(6:30:00 PM) ***DerekNonGeneric breathes a sigh of relief
```